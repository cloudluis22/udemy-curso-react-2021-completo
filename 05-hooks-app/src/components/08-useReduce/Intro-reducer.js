
const initialState = [{
    id: 1,
    todo: 'Comprar la RTX 3060 Ti a buen precio.',
    done: false
}];

const todoReducer = (state = initialState, action) => {

    if (action?.type === 'add') {
        return [...state, action.payload];
    }

    return state;
}

let todos = todoReducer();

const newTodo = {
    id: 2,
    todo: 'Comprar un R5 5600X',
    done: false
}

const addTodoAction = {
    type: 'add',
    payload: newTodo
}

todos = todoReducer(todos, addTodoAction)

console.log(todos);