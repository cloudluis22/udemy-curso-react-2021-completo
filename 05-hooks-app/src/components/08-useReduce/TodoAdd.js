import React from 'react'
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({ handleAddTodo }) => {

    const [{description}, handleInputChange, reset] = useForm({
        description: ''
    });

    const handleSubmit = (evt) => {
        evt.preventDefault();

        if (description.trim().length <= 1) {
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        }

        handleAddTodo(newTodo);
        reset();
    }

    return (
        <div>
            <h4>Agregar TODO</h4>
            <hr />

            <form onSubmit={handleSubmit}>

                <input
                    type="text"
                    name="description"
                    placeholder="TODO..."
                    autoComplete="off"
                    className="form-control"
                    value={description}
                    onChange={handleInputChange}
                />

                <div className="d-grid gap-2">
                    <button className="btn btn-outline-primary mt-1" type="submit">Agregar</button>
                </div>

            </form>
        </div>
    )
}