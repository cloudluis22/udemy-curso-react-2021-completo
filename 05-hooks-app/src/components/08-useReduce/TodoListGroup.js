import React from 'react';
import { TodoListItem } from './TodoListItem';
import PropTypes from 'prop-types';

export const TodoListGroup = ({todos, handleToggle, handleDelete}) => {
    return (
        <ul className="list-group list-group-flush">
            {
                todos.map((todo, i) => {
                   return <TodoListItem  key={todo.id} todo={todo} index={i} handleToggle={handleToggle} handleDelete={handleDelete} />
                })
            }
        </ul>
    )
}

TodoListGroup.propTypes = {
    todos: PropTypes.array.isRequired,
    handleToggle: PropTypes.func.isRequired,
    handleDelete: PropTypes.func.isRequired
}