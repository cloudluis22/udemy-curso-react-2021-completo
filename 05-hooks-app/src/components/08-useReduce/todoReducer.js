
export const todoReducer = (state = [], action) => {

    switch (action.type) {
        case 'add':

            return [...state, action.payload]

        case 'delete':
            return state.filter(todo => todo.id !== action.payload);

        case 'toggle':
            return state.map(todo =>
                (todo.id === action.payload)
                    ? { ...todo, done: !todo.done } // Copia todos los valores anteriores del todo, y done es el valor opuesto al que ten√≠a.
                    : todo // Si no coincide con el payload solo devuelve el todo igual.
            )

        default:
            return state;

    }

}