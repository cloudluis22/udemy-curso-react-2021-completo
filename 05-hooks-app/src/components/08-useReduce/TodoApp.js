import React, { useEffect, useReducer } from 'react';
import { todoReducer } from './todoReducer'
import { TodoListGroup } from './TodoListGroup'
import { TodoAdd } from './TodoAdd';
import './styles.css';

const init = () => {

    let lc = null;

    if (typeof window !== 'undefined') {
        lc = localStorage.getItem('todos');
    }

    // Si hay objetos a parsear los devuelve, si no, solo retorna un arreglo vacÃ­o para el estado inicial.
    return JSON.parse(lc) || [];

}

export const TodoApp = () => {

   const [todos, dispatch] = useReducer(todoReducer, [], init);

    useEffect(() => {
        if (typeof window !== 'undefined') {
            localStorage.setItem('todos', JSON.stringify(todos));
        }
    }, [todos])

    const handleDelete = (todoId) => {
        const action = {
            type: 'delete',
            payload: todoId
        }

        dispatch(action);
    }

    const handleToggle = (todoId) => {

        dispatch({
            type: 'toggle',
            payload: todoId
        })

    }

    const handleAddTodo = (newTodo) => {

        dispatch({
            type: 'add',
            payload: newTodo
        })

    }

    return (
        <div>
            <h1>Todo App ({ todos.length })</h1>
            <hr />

            <div className="row">

                <div className="col-7">
                    <TodoListGroup todos={todos} handleToggle={ handleToggle } handleDelete={ handleDelete } />
                </div>

                <div className="col-5">

                    <TodoAdd handleAddTodo={ handleAddTodo }/>

                </div>

            </div>

        </div>
    )
}
