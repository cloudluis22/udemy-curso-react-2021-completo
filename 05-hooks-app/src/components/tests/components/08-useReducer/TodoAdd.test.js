import React from 'react';
import { shallow } from 'enzyme'
import { TodoAdd } from '../../../08-useReduce/TodoAdd';

describe('Pruebas en el componente de <TodoAdd />', () => {

    const handleAddTodo = jest.fn();

    const wrapper = shallow(<TodoAdd handleAddTodo={ handleAddTodo } />)

    test('Debe de mostrarse correctamente y crear una snapshot.', () => {

        expect(wrapper).toMatchSnapshot();

    });

    test('NO debe de llamar al "handleAddTodo".', () => {

        const formSubmit = wrapper.find('form').prop('onSubmit');
        formSubmit({ preventDefault() { } });

        expect(handleAddTodo).toHaveBeenCalledTimes(0);
    });

    test('Debe de llamar la funciÃ³n de "handleAddTodo".', () => {

        const value = 'Comprar un R5 5600X';
        wrapper.find('input').simulate('change', {
            target: {
                value,
                name: 'description'
            }
        });

        const formSubmit = wrapper.find('form').prop('onSubmit');
        formSubmit({ preventDefault() { } });

        expect(handleAddTodo).toHaveBeenCalledTimes(1);
        expect(handleAddTodo).toHaveBeenCalledWith(expect.any(Object));
        expect(handleAddTodo).toHaveBeenCalledWith({
            id: expect.any(Number),
            desc: value,
            done: false
        });

        // Comprobar que el formulario se reinicia despues del submit.
        expect(wrapper.find('input').prop('value')).toBe('');

    });

});