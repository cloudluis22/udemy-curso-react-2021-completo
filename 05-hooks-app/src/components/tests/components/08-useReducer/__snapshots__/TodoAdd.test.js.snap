// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente de <TodoAdd /> Debe de mostrarse correctamente y crear una snapshot. 1`] = `
<div>
  <h4>
    Agregar TODO
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="TODO..."
      type="text"
      value=""
    />
    <div
      className="d-grid gap-2"
    >
      <button
        className="btn btn-outline-primary mt-1"
        type="submit"
      >
        Agregar
      </button>
    </div>
  </form>
</div>
`;
